aiobotocore/retries/adaptive.py

```diff
@@ -0,0 +1,86 @@
+"""An async reimplementation of the blocking elements from botocore.retries.adaptive."""
+import asyncio
+import logging
+
+from botocore.retries import throttling
+from botocore.retries import standard
+
+# The RateClocker from botocore uses a threading.Lock, but in a single-threaded asyncio
+# program, the lock will be acquired then released by the same coroutine without
+# blocking.
+from botocore.retries.adaptive import RateClocker
+
+from . import bucket
+
+
+logger = logging.getLogger(__name__)
+
+
+def register_retry_handler(client):
+    clock = bucket.Clock()
+    rate_adjustor = throttling.CubicCalculator(starting_max_rate=0,
+                                               start_time=clock.current_time())
+    token_bucket = bucket.AsyncTokenBucket(max_rate=1, clock=clock)
+    rate_clocker = RateClocker(clock)
+    throttling_detector = standard.ThrottlingErrorDetector(
+        retry_event_adapter=standard.RetryEventAdapter(),
+    )
+    limiter = AsyncClientRateLimiter(
+        rate_adjustor=rate_adjustor,
+        rate_clocker=rate_clocker,
+        token_bucket=token_bucket,
+        throttling_detector=throttling_detector,
+        clock=clock,
+    )
+    client.meta.events.register(
+        'before-send', limiter.on_sending_request,
+    )
+    client.meta.events.register(
+        'needs-retry', limiter.on_receiving_response,
+    )
+    return limiter
+
+
+class AsyncClientRateLimiter:
```

Comment:
given you're not inheriting need to patch check the whole class

Reply:
sorry, by patch I meant in test_patches.py make a ref to the class itself, ex:
```python
AsyncClientRateLimiter: {'xxx'},
```

Reply:
no need for any other test changes

Reply:
that way it will has the whole class, so if anything in the class changes it will trigger

Reply:
Do you mean ClientRateLimiter (the botocore class), or AsyncClientRateLimiter (the aiobotocore one?). I added the former in test_patches.py.